wikipedia_transform/wikipedia_transform.py

```diff
@@ -0,0 +1,112 @@
+from wikipedia_transform.models.types import (
+    EmbeddingModelType,
+    EnhancementType,
+    RecordType,
+)
+from wikipedia_transform.readers import WikipediaReader
+
+
+class WikipediaTransform(RecordTransform):
+    """A transformer for Wikipedia articles.
+
+    WikipediaTransform takes in a frozenset of file paths, and uses an embedding model pipeline to convert Wikipedia articles 
+    stored at the file paths into embeddings. When enhancements are set in the environment variables, WikipediaTransform uses a generative AI pipeline to
+    enhance the articles before they are converted into embeddings.
+
+    """
+
+    def __init__(self, file_paths: frozenset[Path]):
+        self.__file_paths: frozenset[Path] = file_paths
+        self.__embedding_pipeline: EmbeddingPipeline | None = (
+            self.__select_embedding_pipeline()
+        )
+
+    def __select_embedding_pipeline(self) -> EmbeddingPipeline | None:
+        """Select and return an embedding model type based on environment variables."""
+
+        if (
+            settings.embedding_model_type == EmbeddingModelType.OPEN_AI
+            and settings.openai_api_key
+        ):
+            return OpenAiEmbeddingPipeline()
+
+        return None
+
+    def __select_generative_ai_pipeline(self) -> GenAiPipeline | None:
+        """Select and return a generative ai model based on environmnet variables."""
+
+        if settings.generative_model_type and settings.openai_api_key:
+            return OpenAiPipeline()
+
+        return None
+
+    def __enhance_articles_with_summaries(
+        self,
+        *,
+        generative_ai_pipeline: GenAiPipeline,
+        wikipedia_articles: tuple[wikipedia.Article, ...],
+        store: StrictBool = True
+    ) -> Iterable[Record]:
+        """Use a generative AI model to enhance Wikipedia articles with summaries."""
+
+        enhanced_wikipedia_articles = [
+            next(
+                iter(
+                    generative_ai_pipeline.enhance_record(
+                        record=wikipedia_article,
+                        record_type=RecordType.WIKIPEDIA,
+                        enhancement_type=EnhancementType.SUMMARY,
+                    )
+                )
+            )
+            for wikipedia_article in wikipedia_articles
+        ]
+
+        if store:
+            summary_store_path = (
+                settings.output_path / "summaries_of_wikipedia_articles.txt"
+            )
+
+            with summary_store_path.open(mode="w") as summary_output:
+                enhanced_wikipedia_articles_as_json = [
+                    json.dumps(
+                        enhanced_wikipedia_article.model_dump(by_alias=True))
+                    for enhanced_wikipedia_article in enhanced_wikipedia_articles
+                ]
+
+                summary_output.writelines(enhanced_wikipedia_articles_as_json)
+
+        yield from enhanced_wikipedia_articles
+
+    def transform(self):
```

Comment:
This code is ~ a pipeline orchestration.

This is a common coding task, and it's good to try doing it yourself once or twice. Beyond that I'd look to frameworks. I like one called Dagster.

Reply:
If you're going to write your own, try to do it in a functional style, where you end up with something like this:

```
load(transform(extract()))
```

Where `extract` yields a sequence of input records, `transform` transforms to a sequence of another type, and `load` saves/loads the latter into some store and returns None (it's pure side-effecting).

Reply:
It may look like:

```
load(transform3(transform2(transform1(extract()))
```

`extract` is not doing any transformation work, but simply pulls data from outside the current process into a raw form.

`load` is not doing any transformation work, but simply serializing and saving transformed data.

Reply:
In the current code:

`extract` raw `Record`s directly into a model, but don't transform
`transform` Record's into embedding `Document`s
`load` Document's into a serialized vector store

Reply:
That said, Dagster doesn't explicitly differentiate between E-T-L phases. Everything is an asset. It's still useful to think in terms of ETL in order to keep transformation code isolated from extraction and loading.